xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -63,33 +61,37 @@
   @Nullable
   private ScheduledHandle ldsRespTimer;
 
-  ServerXdsClient(XdsChannel channel, Node node, SynchronizationContext syncContext,
-      ScheduledExecutorService timeService, BackoffPolicy.Provider backoffPolicyProvider,
-      Supplier<Stopwatch> stopwatchSupplier, boolean newServerApi, String instanceIp) {
+  ServerXdsClient(XdsChannel channel, Node node, ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider, Supplier<Stopwatch> stopwatchSupplier,
+      boolean newServerApi, String instanceIp) {
     super(channel, node, timeService, backoffPolicyProvider, stopwatchSupplier);
-    this.syncContext = checkNotNull(syncContext, "syncContext");
     this.newServerApi = channel.isUseProtocolV3() && newServerApi;
     this.instanceIp = (instanceIp != null ? instanceIp : "0.0.0.0");
   }
 
   @Override
-  void watchListenerData(int port, ListenerWatcher watcher) {
-    checkState(listenerWatcher == null, "ListenerWatcher already registered");
-    listenerWatcher = checkNotNull(watcher, "watcher");
-    checkArgument(port > 0, "port needs to be > 0");
-    this.listenerPort = port;
-    getLogger().log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
-    if (!newServerApi) {
-      updateNodeMetadataForListenerRequest(port);
-    }
-    adjustResourceSubscription(ResourceType.LDS);
-    if (!isInBackoff()) {
-      ldsRespTimer =
-          syncContext
-              .schedule(
-                  new ListenerResourceFetchTimeoutTask(":" + port),
-                  INITIAL_RESOURCE_FETCH_TIMEOUT_SEC, TimeUnit.SECONDS, getTimeService());
-    }
+  void watchListenerData(final int port, final ListenerWatcher watcher) {
+    getSyncContext().execute(new Runnable() {
+      @Override
+      public void run() {
+        checkState(listenerWatcher == null, "ListenerWatcher already registered");
+        listenerWatcher = checkNotNull(watcher, "watcher");
+        checkArgument(port > 0, "port needs to be > 0");
```

Comment:
Shouldn't the above 3 lines (checks) in the main method instead of in the `Runnable.run`? The caller should get the exception instead of the exception being lost in a different thread.

Reply:
Also if `watchListenerData` is called exactly once for a `ServerXdsClient` why even bother with the syncContext here?  When `ServerXdsClient` will be able to support multiple watchers for multiple ports we can then add the syncContext.

Reply:
States meaningful to the task should be checked at the time executing it instead of scheduling it.

Reply:
What? Why not fail immediately if `watcher` is null instead of having it thrown inside `run`? These are all checks on the input arguments to `watchListenerData` so it makes no sense to do them inside the scheduled task.